### PR TITLE
Adding Telemetry headers to public WIL package. These headers add con…

### PR DESCRIPTION
**Pull request by ZenBird**
…venience macros for defining and logging telemetry events. These event can be used for local tracing.

### CODE REVIEW

File: include/wil/Microsoft.Windows.Telemetry.Common.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+#ifndef __MICROSOFT_WINDOWS_TELEMETRY_COMMON_INCLUDED
+#define __MICROSOFT_WINDOWS_TELEMETRY_COMMON_INCLUDED
+
+#include <wil/common.h>
+
+#if defined(__cplusplus) && !defined(__WIL_MIN_KERNEL)
+
+/// @cond
+#ifndef WIL_HIDE_DEPRECATED_1611
```
Comment (dunhor) [inline include/wil/Microsoft.Windows.Telemetry.Common.h:13]: Don't introduce things that are already deprecated; remove these. Seems like this entire header can be deleted.

File: include/wil/Microsoft.Windows.Telemetry.Resource.h [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    WIL_WARN_DEPRECATED_1611_PRAGMA(unique_hheap_ptr)
+
+        // deprecated, use unique_process_heap_ptr instead as that has the correct name.
+        template <typename T>
+    using unique_hheap_ptr = wistd::unique_ptr<T, process_heap_deleter>;
+#endif
+#endif // __WIL_INTERNAL_WINBASE_
+#endif // __WIL_WINBASE_
+
+    // do not add new uses of UNIQUE_STRING_VALUE_EXPERIEMENT, see http://osgvsowi/11252630 for details
```
Comment (dunhor) [inline include/wil/Microsoft.Windows.Telemetry.Resource.h:56]: Seems everything in this header is also either deprecated or has guidance not to use. This file should also be deleted

File: include/wil/Microsoft.Windows.Telemetry.ResultMacros.h [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+#ifndef __MICROSOFT_WINDOWS_TELEMETRY_RESULTMACROS_INCLUDED
+#define __MICROSOFT_WINDOWS_TELEMETRY_RESULTMACROS_INCLUDED
+
+<<<<<<< HEAD
+// Common.h contains logic and configuration macros common to all consumers in Windows and therefore needs to be
+// included before any 'opensource' headers
+#include "Microsoft.Windows.Telemetry.Common.h"
+
+=======
+>>>>>>> e99c787 (Adding Telemetry headers to public WIL package. These headers add convenience macros for defining and logging telemetry events. These event can be used for local tracing.)
```
Comment (ZenBird) [inline include/wil/Microsoft.Windows.Telemetry.ResultMacros.h:16]: Bad merge. Fix.

**Comment by jonwis** [comment]
> #endif

This pattern is a little odd. Maybe instead of forking the macros, use an "expands to nothing" model, like:

```c
#ifdef _GENERIC_PARTB_FIELDS_ENABLED
#define _TLG_GENERIC_PARTB_PARAMS _GENERIC_PARTB_FIELDS_ENABLED,
#else
#define _GENERIC_PARTB_FIELDS_ENABLED
#endif
```

And then use it like this in a single macro:

```c
    #define DEFINE_TRACELOGGING_EVENT_PARAM3_CV(EventId, VarType1, varName1, VarType2, varName2, VarType3, varName3, ...) \
        template<typename T1, typename T2, typename T3> static void EventId(T1 &&varName1, T2 &&varName2, T3 &&varName3, PCSTR correlationVector) \
        { \
            TraceLoggingWrite(TraceLoggingType::Provider(), #EventId, _TLG_GENERIC_PARTB_PARAMS \
                TraceLoggingValue(static_cast<VarType1>(wistd::forward<T1>(varName1)), _wiltlg_STRINGIZE(varName1)), \
                TraceLoggingValue(static_cast<VarType2>(wistd::forward<T2>(varName2)), _wiltlg_STRINGIZE(varName2)), \
                TraceLoggingValue(static_cast<VarType3>(wistd::forward<T3>(varName3)), _wiltlg_STRINGIZE(varName3)), \
                TraceLoggingString(correlationVector, "__TlgCV__"), __VA_ARGS__); \
        }
```

So when the fields are set they're injected, otherwise they aren't

---
Refers to: include/wil/Tracelogging.h:1768 in 3fa3f39. [](commit_id = 3fa3f39c3b4771c21827001fc1fd280311359819, deletion_comment = False)

**Comment by jonwis** [comment]
>                 TraceLoggingValue(static_cast<VarType2>(wistd::forward<T2>(varName2)), _wiltlg_STRINGIZE(varName2)), \

These part-b args should probably come after the caller-provided parameters but before the CV parameter.  Otherwise it makes scripts that use positional-indexed TLG content broken/variable when this is or isn't set.

---
Refers to: include/wil/Tracelogging.h:1734 in 3fa3f39. [](commit_id = 3fa3f39c3b4771c21827001fc1fd280311359819, deletion_comment = False)

**Comment by ZenBird** [comment]
>                 TraceLoggingValue(static_cast<VarType2>(wistd::forward<T2>(varName2)), _wiltlg_STRINGIZE(varName2)), \

Would this apply to the macros without CV as well?

---
In reply to: [871716427](https://github.com/microsoft/wil/pull/190#issuecomment-871716427) [](http://example.com/codeflow?ancestors=871716427,871716427)

---
Refers to: include/wil/Tracelogging.h:1734 in 3fa3f39. [](commit_id = 3fa3f39c3b4771c21827001fc1fd280311359819, deletion_comment = False)

**Comment by ZenBird** [comment]
@MythiliM is added to the review. #Closed

File: README.md [hunk 4]
```diff
@@ -25,6 +25,7 @@ Some things that WIL includes to whet your appetite:
   Preprocessor macros to help you check for errors from Windows API functions,
   in many of the myriad ways those errors are reported, and surface them as
   error codes or C++ exceptions in your code.
+- TODO: Add documentation for telemetry headers
```
Comment (dunhor) [inline README.md:28]: Either address, or remove

**Comment by ZenBird** [comment]
@benkuhn is added to the review. #Closed

**Comment by ZenBird** [comment]
@Wildisc is added to the review. #Closed

File: README.md [hunk 5]
```diff
@@ -25,6 +25,8 @@ Some things that WIL includes to whet your appetite:
   Preprocessor macros to help you check for errors from Windows API functions,
   in many of the myriad ways those errors are reported, and surface them as
   error codes or C++ exceptions in your code.
+- [`include/wil/tracelogging.h`](include/wil/tracelogging.h): This file contains the convenience macros that 
+  enable developers define and log   telemetry. These macros use [`TraceLogging API`](https://docs.microsoft.com/en-us/windows/win32/tracelogging/trace-logging-portal) to log data. This data can be viewed in tools such as [`Windows Performance Analyzer`](https://docs.microsoft.com/en-us/windows-hardware/test/wpt/windows-performance-analyzer).
```
Comment (jonwis) [inline README.md:29]: Please word-wrap this block like the others.

**Review by jonwis — CHANGES_REQUESTED** [review]

**Review by jonwis — APPROVED** [review]

**Pull request merged**